### PR TITLE
fix: fix boolean return type mismatch in C ABI wrapper

### PR DESCRIPTION
**Pull request by kaeun97**
Fixes https://github.com/NVIDIA/numba-cuda/issues/157.

### CODE REVIEW

File: numba_cuda/numba/cuda/target.py [hunk 1]
```diff
@@ -453,6 +453,21 @@ def _make_call_helper(self, builder):
...
 
     def return_value(self, builder, retval):
+        expected_type = builder.function.ftype.return_type
+        actual_type = retval.type
+
+        # If types don't match, we need to cast
+        if actual_type != expected_type:
+            if isinstance(actual_type, ir.IntType) and isinstance(
+                expected_type, ir.IntType
+            ):
```
Comment (kaeun97) [inline numba_cuda/numba/cuda/target.py]: Should we consider other types as well?

Comment (kaeun97) [reply numba_cuda/numba/cuda/target.py]: Or throw an error if there is an unexpected type mismatch?

**Comment by gmarkall** [comment]
/ok to test https://github.com/NVIDIA/numba-cuda/pull/770/commits/3ec7629fae2aeced755325a93931fc4c8c3c4296

**Comment by gmarkall** [comment]
Will also check this locally with #717 to see how they interact.

**Comment by gmarkall** [comment]
> Will also check this locally with #717 to see how they interact.

With this merged into #717 (and fixed up conflicts) all tests still pass.

File: numba_cuda/numba/cuda/target.py (hunk 1 above)
Comment (gmarkall) [reply numba_cuda/numba/cuda/target.py]: I'm hoping the `i1` boolean / `i8` return type is the only edge case we need to handle.

For some background thoughts: it's a bit fiddly because usually if you're trying to accept as an argument or return something that's not quite the right type, you would be able use `CUDATargetContext.cast()` (i.e. `self.context.cast()` just here) to use the target context's mechanism to handle any case appropriately. But it needs to be told the Numba type of the types to cast from / to, and we don't have them here.

I'm inclined to go with this fix as-is, but if we start butting up against other edge cases, perhaps we should think about how we could use `context.cast()` instead of further customisation here.

**Review by gmarkall — APPROVED** [review]
The change looks good, it just needs a merge of `main` and the conflict resolving.

Because of #717:
* `CUDACABICallConv` moved to `numba_cuda/numba/cuda/core/callconv.py`.
* `cabi_wrap_function` was deleted (so that part of the fix shouldn't be needed anymore). `cabi_wrap_function` was a hack to handle the fact we couldn't compile with different calling conventions within one compilation, but #717 removed that limitation.

I did the merge the other way when testing, but if it helps to see how I resolved conflicts, I pushed that up here: https://github.com/gmarkall/numba-cuda/tree/pr-717-770 ([merge commit](https://github.com/gmarkall/numba-cuda/commit/3bbda4dec72e137fd0ac6184bd45ae7c3b6224d6))

**Comment by gmarkall** [comment]
/ok to test https://github.com/NVIDIA/numba-cuda/pull/770/commits/c099a7c3ff49a5a67b6bd65ac0fd1c1ab104ea77

**Comment by gmarkall** [comment]
@kaeun97 I hope you don't mind, I did the merge of `main` because we're hoping to get this merged then make a release incorporating it.

**Comment by kaeun97** [comment]
Thanks for stepping in @gmarkall ! I was planning to do it tonight - was not aware that this was urgent. Appreciate it!
